File: leetcode/rotate-array.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        if (nums.size() == 1)
            return;
        
        if (k > nums.size())
            k = k%nums.size();
        
        
        vector<int> answer = vector<int>(k);
        
        auto it = nums.end()-1;
        
        answer.insert(answer.begin() + k, nums.begin(), nums.end() - k);
        
        for(int i = k-1; i >= 0; --i) {
            answer[i] = *it;
            --it;
        }
        
        nums = answer;
    }
};

File: leetcode/rotate-array.py
from typing import List


class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        d = {}
        length = len(nums)

        for i in range(length):
            new_index = (i + k) % length

            if new_index not in d:
                d[new_index] = nums[new_index]

            nums[new_index] = d[i] if i in d else nums[i]
